core/utils/codel.h

```diff
@@ -0,0 +1,267 @@
+      : delay_target_(target),
+        window_(window),
+        time_above_target_(0),
+        time_till_drop_(MicroSecondTime() + window),
+        drop_count_(0),
+        dropping_(0),
+        queue_(init_queue),
+        drop_func_(drop_func) {}
+
+  virtual ~Codel() {
+    Wrapper* w;
+    while ((w = queue_->Pop())) {
+      Drop(w);
+    }
+    delete queue_;
+  }
+
+  // Enqueue one object. Takes a pointer to the object to be added. Returns 0 on
+  // success.
+  int Push(T* ptr) {
+    Wrapper* w = new Wrapper(ptr, MicroSecondTime());
+    int err = queue_->Push(w);
+    if (err) {
+      delete w;
+    }
+    return err;
+  }
+
+  // Enqueue multiple objects. Takes a pointer an table of object pointers, the
+  // number of objects to be added. Returns 0 on success.
+  int Push(T** ptr, size_t count) {
+    if (Capacity() - Size() <= count) {
+      return -2;
+    }
+
+    for (size_t i = 0; i < count; i++) {
+      if (int err = Push(ptr[i])) {
+        return err;
+      }
+    }
+    return 0;
+  }
+
+  // Retrieves the next entry from the queue and in the process, potentially drops
+  // pointers as well as changes between dropping state and not dropping state.
+  // Returns the pointer to the dequeued object if there is an object to
+  // dequeue. Otherwise, returns nullptr.
+  T* Pop() {
+    Wrapper* w = RingDequeue();
+    if (!w) {
+      dropping_ = 0;
+      return nullptr;
+    }
+
+    uint64_t now = MicroSecondTime();
+    if (dropping_) {
+      // if in dropping state, drop pointers until next drop time is greater
+      // than
+      // the current time.
+      w = DropDequeue(w);
+
+    } else if (w->drop && ((now - time_till_drop_ < window_) ||
+                           (now - time_above_target_ >= window_))) {
+      // if not in dropping state, determine whether to enter drop state and if
+      // so, drop current packet, get a new packet and reset the drop counter.
+      Drop(w);
+      w = RingDequeue();
+
+      if (w) {
+        dropping_ = 1;
+        if (now - time_till_drop_ < window_ && drop_count_ > 2) {
+          drop_count_ -= 2;
+        } else {
+          drop_count_ = 1;
+        }
+        time_till_drop_ = NextDrop(now);
+      }
+    }
+
+    T* ptr;
+    if (w) {
+      ptr = w->ptr;
+      w->ptr = nullptr;
+      Drop(w);
+    } else {
+      ptr = nullptr;
+    }
+    return ptr;
+  }
+
+  // Retrieves the next count entries from the queue and in the process, potentially
+  // drops pointers as well as changes between dropping state and not dropping state.
+  // Takes the number of objects to be dequeued. Returns a pointer to a table of
+  // pointers (objects) that has been filled if there are count objects.
+  // Otherwise, returns nullptr.
+  T** Pop(size_t count) {
+    if (Size() < count) {
+      return nullptr;
+    }
+
+    T** objs = new T*[count];
+    for (size_t i = 0; i < count; i++) {
+      T* next_obj = Pop();
+      if (!next_obj) {
```

Comment:
ah, this was a hold over code before I added the check at the beginning of the function to make it atomic. it will just return a nullptr if there are not enough items to dequeue and will never touch this code. I deleted it.